sycl/plugins/esimd_emulator/pi_esimd_emulator.hpp

```diff
@@ -104,48 +112,78 @@ struct _pi_mem : _pi_object {
 
   pi_context Context;
 
+  // To be used for piEnqueueMemBufferMap
   char *MapHostPtr = nullptr;
 
-  // Mutex for load/store accessing
+  // Mutex for load/store accessing from memory instrinsic
+  // implementations
   std::mutex mutexLock;
 
-  // Surface index used by CM
-  int SurfaceIndex;
+  // Surface index
+  unsigned int SurfaceIndex;
 
   virtual ~_pi_mem() = default;
 
   _pi_mem_type getMemType() const { return MemType; };
 
+  bool isHostProvidedMem() const { return isHostMem; };
+
+  pi_result addMapping(void *MappedTo);
+  pi_result removeMapping(void *MappedTo);
+
 protected:
   _pi_mem(pi_context ctxt, char *HostPtr, _pi_mem_type MemTypeArg,
-          int SurfaceIdxArg)
-      : Context{ctxt}, MapHostPtr{HostPtr},
-        SurfaceIndex{SurfaceIdxArg}, MemType{MemTypeArg} {}
+          unsigned int SurfaceIdxArg, bool isHostMemArg)
+      : Context{ctxt}, MapHostPtr{HostPtr}, SurfaceIndex{SurfaceIdxArg},
+        MemType{MemTypeArg}, isHostMem{isHostMemArg} {}
 
 private:
   _pi_mem_type MemType;
+
+  bool isHostMem;
+
+  // TODO: Runtime guarantees LIFO across multiple threads for
+  // piEnqueueMemBufferMap/Unmap? No duplicated mappings for same
+  // (base, offset)?
+  std::stack<void *> Mappings;
 };
 
+typedef union _cm_buffer_ptr {
+  // For normal buffer
+  cm_support::CmBuffer *GPUPtr;
```

Comment:
```suggestion
  cm_support::CmBuffer *GPUPtr = nullptr;
```

Reply:
my suggestion is to keep info about how to interpret the contents of this union encapsulated within a single object:

struct cm_buffer_ptr_slot {
  enum type {
    type_none, type_regular, type_user_provided
  };
  type tag = type_none;
  cm_support::CmBuffer *RegularBufPtr;
  cm_support::CmBufferUP *UPBufPtr;

  cm_support::CmBuffer* getBuffer() {
    assert(tag == type_regular);
    return RegularBufPtr;
  }

  cm_support::CmBufferUP* getBufferUP() {
    assert(tag == type_user_provided);
    return UPBufPtr;
  }
}

Reply:
Done.